dialogue-core/src/main/java/com/palantir/dialogue/core/RoundRobinChannel.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.core;
+
+
+import com.google.common.util.concurrent.ListenableFuture;
+import com.palantir.dialogue.Endpoint;
+import com.palantir.dialogue.Request;
+import com.palantir.dialogue.Response;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Round robins requests across many channels, attempting to choose a channel that has some available capacity.
+ */
+public final class RoundRobinChannel implements LimitedChannel {
+
+    private final AtomicInteger currentHost = new AtomicInteger(0);
+    private final List<LimitedChannel> delegates;
+
+    public RoundRobinChannel(List<LimitedChannel> delegates) {
```

Comment:
I think we're missing a concept of a `ChannelSelector`. We will need to force rotation when requests fail in certain ways, but this is only aware of pre-flight limiting, and doesn't seem like the right place for exceptional response interpretation.

Reply:
Think you just end up with another implementation of the `LimitedChannel` which returns Optional.empty() for some period of time after a particular exception. The goal here is to back each of these things into very separate layers to avoid entangling things like they are today in remoting